### PR TITLE
feat(ext/ffi): Safe number pointers

### PR DESCRIPTION
**Pull request by aapoalas**
This PR changes the JS representation of pointers returned by the FFI API to use plain JS numbers on 64 bit systems when the pointer value can be safely represented as a JS number. (Note: There would be another option as well, which is to always turn pointers into f64 as they are, bitwise, and save that into a JS `Number`. Whether or not that actually works depends entirely on if V8 does any kind of changes on the f64 it's given, eg. unifying any NaN values to be the same NaN.)

### CODE REVIEW

**Review by littledivy — COMMENTED** [review]
Please add typescript tests in `test_ffi/tests/ffi_types.ts`

**Comment by aapoalas** [comment]
> Please add typescript tests in `test_ffi/tests/ffi_types.ts`

Turns out these are quite well handled by the existing type tests. The tests had just turned off with `deno check` no longer type checking by default.

File: ext/ffi/jit_trampoline.rs [hunk 1]
```diff
@@ -22,12 +22,14 @@ fn native_arg_to_c(ty: &NativeType) -> &'static str {
...
-    NativeType::I64 | NativeType::ISize => "int64_t",
     NativeType::Void => "void",
     NativeType::F32 => "float",
     NativeType::F64 => "double",
-    _ => unimplemented!(),
+    NativeType::U64 => "uint64_t",
+    NativeType::I64 => "int64_t",
+    NativeType::ISize => "intptr_t",
+    NativeType::USize => "uintptr_t",
+    NativeType::Pointer | NativeType::Function => "void*",
```
[the remark this answers is not in the document]
Comment (aapoalas) [reply ext/ffi/jit_trampoline.rs:32]: I opened a new issue for this: https://github.com/denoland/deno/issues/15292

It seems that the main case where this is true is Harvard architecture, which is not supported by Deno either. These seem to usually be lower-bitness environments, eg. 16 bit data pointers, which would mean that compiling V8 for those might not be possible in the first place. Compiling Rust might not work either.

[the remark this answers is not in the document]
Comment (Phosra) [reply ext/ffi/jit_trampoline.rs:32]: > I opened a new issue for this: https://github.com/denoland/deno/issues/15292
> 
> It seems that the main case where this is true is Harvard architecture, which is not supported by Deno either.

> These seem to usually be lower-bitness environments, eg. 16 bit data pointers, which would mean that compiling V8 for those might not be possible in the first place.

Indeed, this is the usual case, that `sizeof::<fn()>() < size_of::<usize>()`.

> Compiling Rust might not work either.

I believe Rust could still work on such platforms, as it requests passing `fn()` as FFI functions.
<https://doc.rust-lang.org/std/primitive.fn.html>
And Rust acknowledges that function pointers may not be pointer-sized.
<https://rust-lang.github.io/unsafe-code-guidelines/layout/function-pointers.html>

But if we have confirmation that V8 can't support such platforms, then it is perfectly understandable to disregard this situation as not relevant.
(I'm not sure if V8's JiT compiler would work on such platforms, as Harvard platforms are more complicated to do so on.)

[the remark this answers is not in the document]
Comment (littledivy) [reply ext/ffi/jit_trampoline.rs:32]: Perhaps a const assert `assert!(sizeof::<fn()>() == size_of::<usize>())` is fine.

**Review by littledivy — APPROVED** [review]
LGTM!

**Pull request merged**